pkg/validate/scoped/hydrate/namespace_hydrator.go

```diff
@@ -15,85 +15,148 @@
+				if obj.GetNamespace() == "" {
+					if objs.Scope == declared.RootReconciler {
+						obj.SetNamespace(metav1.NamespaceDefault)
+					} else {
+						obj.SetNamespace(string(objs.Scope))
+					}
+				}
+				result = append(result, obj)
 			}
-		} else {
-			if obj.GetNamespace() == "" {
-				obj.SetNamespace(objs.DefaultNamespace)
-			}
-			result = append(result, obj)
 		}
+		if errs != nil {
+			return errs
+		}
+		objs.Namespace = result
+		return nil
 	}
-	if errs != nil {
-		return errs
-	}
-	objs.Namespace = result
-	return nil
 }
 
 // buildSelectorMap processes the given cluster-scoped objects to return a map
 // of NamespaceSelector names to the namespaces that are selected by each one.
 // Note that this modifies the Scoped objects to filter out the
 // NamespaceSelectors since they are no longer needed after this point.
-func buildSelectorMap(objs *objects.Scoped) (map[string][]string, status.MultiError) {
-	var namespaces, nsSelectors, others []ast.FileObject
+func buildSelectorMap(ctx context.Context, c client.Client, objs *objects.Scoped) (map[string][]string, status.MultiError) {
+	var namespaces, others []ast.FileObject
+	nsSelectorMap := make(map[string]labels.Selector)
+	selectedNamespaceMap := make(map[string][]labels.Selector)
+	var errs status.MultiError
+	hasDynamicNSSelector := false
+
 	for _, obj := range objs.Cluster {
 		switch obj.GetObjectKind().GroupVersionKind() {
 		case kinds.Namespace():
 			namespaces = append(namespaces, obj)
 		case kinds.NamespaceSelector():
-			nsSelectors = append(nsSelectors, obj)
+			nsSelector, err := convertToNamespaceSelector(obj)
+			if err != nil {
+				errs = status.Append(errs, err)
+				continue
+			}
+			selector, err := labelSelector(nsSelector)
+			if err != nil {
+				errs = status.Append(errs, err)
+				continue
+			}
+			nsSelectorMap[nsSelector.Name] = selector
+
+			if nsSelector.Spec.Mode == v1.NSSelectorDynamicMode {
+				hasDynamicNSSelector = true
+			}
 		default:
 			others = append(others, obj)
 		}
 	}
 
-	var errs status.MultiError
-	selectorMap := make(map[string][]string)
-
-	for _, obj := range nsSelectors {
-		var selected []string
-		selector, err := labelSelector(obj)
-		if err != nil {
+	if objs.AllowAPICall && objs.DynamicNSSelectorEnabled != hasDynamicNSSelector {
+		// The DynamicNSSelectorEnabled state mismatches with the required state, set the
+		// annotation so that the reconciler-manager will recreate the reconciler
+		// with the correct setting for the Namespace watch.
+		if err := setDynamicNSSelectorEnabled(ctx, c, objs.SyncName, hasDynamicNSSelector); err != nil {
 			errs = status.Append(errs, err)
-			continue
+			return nil, errs
 		}
+	}
 
+	nsList := &corev1.NamespaceList{}
+	// Fetch on-cluster Namespaces if allowing API calls and hasDynamicNSSelector,
+	// so that the NamespaceSelector can select resources matching on-cluster Namespaces.
+	if objs.AllowAPICall && hasDynamicNSSelector {
+		if err := c.List(ctx, nsList); err != nil {
```

Comment:
Instead of listing all namespaces and then filtering by labels in code, can we use label selectors on the list call?

Reply:
There might be multiple NamespaceSelectors. If we use label selector, that means multiple API calls or each label selector.
Without using label selector, there is only one single API call.

Reply:
Is it talking to the API server? With namespace watches can it read from cache instead of calling the API server?

Reply:
Yes, it talks to the API server. The namespace watch is in the  namespace controller within the controller-manager running in a separate goroutine.  This namespace_hydrator is in the reconciler thread, so it won't be able to access the cache.

Reply:
Ah, good point. I'm a bit concerned about directly calling the API server. Each reconciler could be listing all namespaces from the API server, and that's pretty expensive.

Reply:
Also considered storing all namespaces in the memory (namespace controller's state) with the namespace controller's cache, and let the reconciler thread read from the namespaces from the state. However, found https://medium.com/@timebertt/kubernetes-controllers-at-scale-clients-caches-conflicts-patches-explained-aa0f7a8b4332#:~:text=Controllers%20must%20not,should%20be%20improved.

> Controllers must not use the same in-memory object concurrently in multiple goroutines. For example, decoding results from the API server in multiple goroutines into the same maps (e.g., labels, annotations) will cause panics because of concurrent map writes. Also, reading from an in-memory API object in one goroutine while decoding into it in another goroutine will yield non-atomic reads, meaning data might be corrupt and represent an invalid/non-existing API object.
Therefore, if you need to use the same in-memory object in multiple goroutines concurrently (e.g., shared state), remember to leverage proper synchronization techniques like channels, mutexes, atomic.Value and/or copy the object prior to use. The average controller however, will not need to share in-memory API objects between goroutines, and it's typically an indicator that the controller's design should be improved.

It feels like using the cache requires to refactor the reconciler thread to use the controller-runtime library, which sets up the watch and uses the cache in the same thread.

Reply:
Let's enhance that later as a follow-up